### ISSUE TITLE
 error with absence of oauth:twitter key

### ISSUE THREAD

**Issue by vetsinen-onlyfans** [superfaceai/passport-twitter-oauth2#36]
can you say what is a reason for error with absent  key oauth:twitter on callback?
debug shows error in if 
node_modules/passport-oauth2/lib/state/pkcesession.js
line 67
  var key = this._key;
  if (!req.session[key]) {
    return callback(null, false, { message: 'Unable to verify authorization request state.' });
  }

**Comment by vetsinen-onlyfans** [comment]
> Hello, I'm not sure I can help you without seeing how you setup the strategy and Express. Do you have `express-session` set up correctly?

module.exports = new TwitterStrategy({
    passReqToCallback: true,
    callbackURL: config.twitter.login.callbackURL,

    // consumerKey: config.twitter.login.consumerKey,
    // consumerSecret: config.twitter.login.consumerSecret,

    clientID: config.twitter.login.clientID,
    clientSecret: config.twitter.login.clientSecret,
    clientType: 'public',

}, (req, accessToken, refreshToken, params, done) => {
    console.log('twitter ', params)
    return social_auth(req, params, done, 'twitter');
})

**Comment by vetsinen-onlyfans** [comment]
> Hello, I'm not sure I can help you without seeing how you setup the strategy and Express. Do you have `express-session` set up correctly?

`
module.exports = new TwitterStrategy({
    passReqToCallback: true,
    callbackURL: config.twitter.login.callbackURL,

    // consumerKey: config.twitter.login.consumerKey,
    // consumerSecret: config.twitter.login.consumerSecret,

    clientID: config.twitter.login.clientID,
    clientSecret: config.twitter.login.clientSecret,
    clientType: 'public',

}, (req, accessToken, refreshToken, params, done) => {
    console.log('twitter ', params)
    return social_auth(req, params, done, 'twitter');
})
`

**Comment by jnv** [comment]
I don't think it makes sense to combine the `public` client with client secret, but to be honest I never tried to implement the public client. Did you, @janhalama?

**Comment by janhalama** [comment]
I tried the `public` client. You do not need to provide client secret.

This is the Strategy setup:
```
new TwitterStrategy(
    {
      clientID: process.env.TWITTER_CLIENT_ID,
      callbackURL: `${process.env.BASE_URL}/auth/twitter/callback`,
      clientType: 'public',
    },
    (accessToken, refreshToken, profile, done) => {
      onAuthSuccess({ accessToken, refreshToken });
      return done(null, {
        displayName: profile.displayName,
      });
    }
  )
```

What is breaking the authentication is the passReqToCallback option enabled. I will look into int tomorrow.

**Comment by janhalama** [comment]
I checked it once again and the problem was in verification callback implementation. The strategy works just fine with the option passReqToCallback enabled.

Here is the Strategy setup:
```
new TwitterStrategy(
    {
      passReqToCallback: true,
      clientID: process.env.TWITTER_CLIENT_ID,
      callbackURL: `${process.env.BASE_URL}/auth/twitter/callback`,
      clientType: 'public',
    },
    (_req, accessToken, refreshToken, profile, done) => {
      onAuthSuccess({ accessToken, refreshToken });
      return done(null, {
        displayName: profile.displayName,
      });
    }
  )
```

I tested it with the [twitter-demo example](https://github.com/superfaceai/twitter-demo). The only change to the project is the code above.

**Comment by janhalama** [comment]
From the error you provided it looks like some problem with express-session set up. @jnv already pointed it out.

**Comment by jnv** [comment]
I think I should mention session requirement + possible way how to work around it in the readme, see also #23. Probably add FAQ question to collect these.
